dbms/src/Interpreters/NullAwareSemiJoinHelper.cpp

```diff
@@ -0,0 +1,470 @@
+    runStep<NASemiJoinStep::NOT_NULL_KEY_CHECK_NULL_ROWS>(res_list, next_step_res_list);
+    res_list.swap(next_step_res_list);
+    if (res_list.empty())
+        return;
+
+    runStep<NASemiJoinStep::NULL_KEY_CHECK_NULL_ROWS>(res_list, next_step_res_list);
+    res_list.swap(next_step_res_list);
+    if (res_list.empty())
+        return;
+
+    runStepAllBlocks(res_list);
+}
+
+template <ASTTableJoin::Kind KIND, ASTTableJoin::Strictness STRICTNESS, typename Mapped>
+template <NASemiJoinStep STEP>
+void NASemiJoinHelper<KIND, STRICTNESS, Mapped>::runStep(std::list<NASemiJoinHelper::Result *> & res_list, std::list<NASemiJoinHelper::Result *> & next_res_list)
+{
+    static_assert(STEP == NASemiJoinStep::NOT_NULL_KEY_CHECK_MATCHED_ROWS || STEP == NASemiJoinStep::NOT_NULL_KEY_CHECK_NULL_ROWS || STEP == NASemiJoinStep::NULL_KEY_CHECK_NULL_ROWS);
+
+    auto it = res_list.begin();
+    while (it != res_list.end())
+    {
+        if ((*it)->getStep() != STEP)
+        {
+            next_res_list.emplace_back(*it);
+            it = res_list.erase(it);
+            continue;
+        }
+        ++it;
+    }
+
+    std::vector<size_t> offsets;
+    size_t block_columns = block.columns();
+    Block exec_block = block.cloneEmpty();
+
+    while (!res_list.empty())
+    {
+        MutableColumns columns(block_columns);
+        for (size_t i = 0; i < block_columns; ++i)
+        {
+            /// Reuse the column to avoid memory allocation.
+            /// Correctness depends on the fact that equal and other condition expressions do not
+            /// removed any column, namely, the columns will not out of order.
+            /// TODO: Maybe we can reuse the memory of new columns added by expressions?
+            columns[i] = exec_block.safeGetByPosition(i).column->assumeMutable();
+            columns[i]->popBack(columns[i]->size());
+        }
+
+        size_t max_pace = std::max(1, max_block_size / res_list.size());
+        size_t current_offset = 0;
+        offsets.clear();
+
+        for (auto & res : res_list)
+        {
+            size_t prev_offset = current_offset;
+            res->template fillRightColumns<Mapped, STEP>(columns, left_columns, right_columns, null_rows, current_offset, max_pace);
+
+            /// Note that current_offset - prev_offset may be zero.
+            if (current_offset > prev_offset)
+            {
+                for (size_t i = 0; i < left_columns; ++i)
+                    columns[i]->insertManyFrom(*block.getByPosition(i).column.get(), res->getRowNum(), current_offset - prev_offset);
+            }
+
+            offsets.push_back(current_offset);
+            if (current_offset >= max_block_size)
+                break;
+        }
+
+        /// Move the columns to exec_block.
+        /// Note that this can remove the new columns that are added in the last loop
+        /// from equal and other condition expressions.
+        exec_block = block.cloneWithColumns(std::move(columns));
```

Comment:
I find it can not use `setColumns` because `exec_block` may have some new columns after executing the expressions. Using `cloneWithColumns` can remove these new columns.